### PR TITLE
SI-8835 Lazier slice for Iterator

### PR DESCRIPTION
**Pull request by som-snytt**
An iterator for slicing that does not drop eagerly.

A series of `take m drop n` that results in nothing taken
will not call the underlying iterator.

Chained invocations do not create intermediate iterators.

Adds eager `drop` to `IterableSplitter`, which overrides `take` and `slice`.

Let `stringOf` handle `ParIterable` so your ScalaCheck labels can be shown truncated, at least during debug.

This PR against 2.12 (because of the change to laziness)
includes the expanded unit test from #4075, but does not
delete the redundant test files.

This is similar to a proposal at #3963: just track drop and take numbers (skip and limit counters) until forced.

There may be a question whether this is worth it, in the space between deprecated views and the collections getting rewritten by multiple parties.

The other question is, if `take` and `drop` and `slice` are defined in terms of a template method, should they be final?

Review by @Ichoran

### CODE REVIEW

**Comment by som-snytt** [comment]
PLS REBUILD ALL

**Comment by som-snytt** [comment]
A parallel collections test is hung, but that problem predates this PR, see #4082.  Apparently the test doesn't always hang, which is the true beauty of parallelism: there's always a parallel universe in which your test has run to completion successfully.

**Comment by som-snytt** [comment]
Now that the pc test runs, it doesn't like the lazy drop business.  I wonder if it's a test artifact, but I'll have to understand the test first. Edit: looks like the pc iterable splitter iterator mechanism overrides slice and dice, uh, take but not drop.  That's my guess after learning a bit how scalacheck works.  A strange artifact was that the test was handed Nil to compare to a non-empty ParVector; possibly that occurred during minimization, but it makes the test rig look broken.

**Comment by som-snytt** [comment]
PLS REBUILD ALL

**Comment by lrytz** [comment]
ping @Ichoran. To avoid conflicts, we should first merge #4075, then merge 2.11.x into 2.12.x, and rebase this.

**Comment by som-snytt** [comment]
PLS REBUILD ALL

File: src/library/scala/collection/Iterator.scala [hunk 1]
```diff
@@ -191,11 +191,56 @@ object Iterator {
...
+        } else
+          dropping = 0
+      }
+    def hasNext = { skip(); remaining != 0 && underlying.hasNext }
+    def next()  = {
+      skip()
+      if (remaining > 0) {
+        remaining -= 1
+        underlying.next()
+      } else if (unbounded) {
```
Comment (Ichoran) [inline src/library/scala/collection/Iterator.scala]: This would be easier to read if the single-line statements were on the same line as the conditions.

File: src/library/scala/collection/Iterator.scala [hunk 2]
```diff
@@ -191,11 +191,56 @@ object Iterator {
...
+        underlying.next()
+      } else {
+        empty.next()
+      }
+    }
+    override protected def sliceIterator(from: Int, until: Int): Iterator[A] = {
+      val lo = from max 0
+      def adjustedBound = if (unbounded) -1 else remaining - lo max 0
```
Comment (Ichoran) [inline src/library/scala/collection/Iterator.scala]: I realize `-` has higher prececence than `max`, but it doesn't really read that way, so parens would be helpful.

File: src/library/scala/collection/Iterator.scala [hunk 3]
```diff
@@ -191,11 +191,56 @@ object Iterator {
...
+      val rest =
+        if (until < 0) adjustedBound          // respect current bound, if any
+        else if (until <= lo) 0               // empty
+        else if (unbounded) until - lo        // now finite
+        else adjustedBound min (until - lo)   // keep lesser bound
+      if (rest == 0) empty
+      //else if (lo == 0 && rest < 0) this
```
Comment (Ichoran) [inline src/library/scala/collection/Iterator.scala]: No reason to have this line any more is there?  (Is it there to remind us to do performance testing both ways?)

File: src/library/scala/collection/Iterator.scala [hunk 4]
```diff
@@ -191,11 +191,56 @@ object Iterator {
...
+      else {
+        dropping += lo
+        remaining = rest
+        this     //new SliceIterator(underlying, dropping + lo, rest)
```
Comment (Ichoran) [inline src/library/scala/collection/Iterator.scala]: Do you mean "Equivalent to new ..." or was this an alternate implementation?

File: test/junit/scala/collection/IteratorTest.scala [hunk 5]
```diff
@@ -25,4 +26,147 @@ class IteratorTest {
...
+    val xs1 = it1 take 10
+    assertEquals(10, xs1.size)
+  }
+  @Test def foreach(): Unit = {
+    val it1 = Iterator.from(0) take 20
+    var n = 0
+    it1 foreach { n += _ }
+    assertEquals(190, n)
+  }
+  /* ???
```
Comment (Ichoran) [inline test/junit/scala/collection/IteratorTest.scala]: I wholeheartedly agree: ???.  Why is this here?

File: test/junit/scala/collection/IteratorTest.scala [hunk 6]
```diff
@@ -25,4 +26,147 @@ class IteratorTest {
...
+  @Test def range2(): Unit = {
+    val r1start = 0
+    val r1end = 10
+    val r1step = 1
+    val r1 = Iterator.range(r1start, r1end, r1step)
+    val r2 = Iterator.range(r1start, r1end, r1step + 1)
+    val r3 = Iterator.range(r1end, r1start, -r1step)
+    val r4 = Iterator.range(0, 10, 11)
+    // 10 + 5 + 10 + 1
+    assertEquals(10 + 5 + 10 + 1, r1.size + r2.size + r3.size + r4.size)
```
Comment (Ichoran) [inline test/junit/scala/collection/IteratorTest.scala]: These magic values (10 + 5 + 10 + 1) depend on r1start, r1end, and r1step.  I'm not sure it makes sense to have this mix of vals and literals.

File: src/library/scala/collection/Iterator.scala (hunk 4 above)
Comment (som-snytt) [inline src/library/scala/collection/Iterator.scala]: I'll zap the extraneous comments.  It means, "In case you were about to say, did you consider this?, the answer is yes, we're specifically not doing that."  But it's not worth adding words to that effect.

File: src/library/scala/collection/Iterator.scala (hunk 1 above)
Comment (som-snytt) [inline src/library/scala/collection/Iterator.scala]: Agree.  I've been writing java code at the office where checkstyle maximizes lines consisting only of close braces.

File: test/junit/scala/collection/IteratorTest.scala [hunk 7]
```diff
@@ -25,4 +26,147 @@ class IteratorTest {
...
+  }
+  /* ???
+  @Test def forall(): Unit = {
+    val it1 = Iterator.from(0)
+    val it2 = Iterator.from(1)
+    0
+  }
+  */
+  // ticket #429
+  @Test def fromArray(): Unit = {
```
Comment (Ichoran) [inline test/junit/scala/collection/IteratorTest.scala]: Collections tests will get all sorts of stuff like this; even if it doesn't now, this sort of test will be much more comprehensively performed there.  (Not to say it can't stay, but it hopefully won't serve much point.)

File: test/junit/scala/collection/IteratorTest.scala [hunk 8]
```diff
@@ -25,4 +26,147 @@ class IteratorTest {
...
+    var xs1 = a.slice(0, 1).iterator
+    var xs2 = a.slice(0, 2).iterator
+    var xs3 = a.slice(0, 3).iterator
+    var xs4 = a.slice(0, 4).iterator
+    assertEquals(14, xs0.size + xs1.size + xs2.size + xs3.size + xs4.size)
+  }
+  @Test def toSeq(): Unit = {
+    assertEquals("1x2x3x4x5", List(1, 2, 3, 4, 5).iterator.mkString("x"))
+  }
+  @Test def indexOf(): Unit = {
```
Comment (Ichoran) [inline test/junit/scala/collection/IteratorTest.scala]: Where is this from?  Seems really out of place here.

File: src/library/scala/collection/Iterator.scala (hunk 2 above)
Comment (som-snytt) [inline src/library/scala/collection/Iterator.scala]: Ha, also agree. Checkstyle complains about unnecessary parens in java...  There's also that two sets of parens are needed, `if (c) r else ((x - y) max 0)` because `else (x-y) max 0` also makes you parse it mentally.  My theory was that no parens means "it does what must make sense", but that doesn't mean much if the sense isn't obvious.

**Comment by Ichoran** [comment]
Looks pretty good.  I think the underlying logic is sound and I didn't see any obvious corner cases that weren't being caught; just a bit of code-tidying and it should be good to go.

File: test/junit/scala/collection/IteratorTest.scala (hunk 5 above)
Comment (som-snytt) [inline test/junit/scala/collection/IteratorTest.scala]: The source of the refactor was ambiguous/tested nothing.  I'll look again to see if there's anything to recover.

File: test/junit/scala/collection/IteratorTest.scala (hunk 6 above)
Comment (som-snytt) [inline test/junit/scala/collection/IteratorTest.scala]: I didn't remember anymore. The iterators.scala test had

```
    // 10 + 5 + 10 + 1
    r1.length + r2.length + r3.length + r4.length
```

and

```
check_success("check_range2",   check_range2,    26)
```

I'll work on my algebra.

File: test/junit/scala/collection/IteratorTest.scala (hunk 7 above)
Comment (som-snytt) [inline test/junit/scala/collection/IteratorTest.scala]: I hope so too.

File: test/junit/scala/collection/IteratorTest.scala (hunk 8 above)
Comment (som-snytt) [inline test/junit/scala/collection/IteratorTest.scala]: Prehistory, meaning pre-2.8.  Did that ever test iterator.indexOf?

https://github.com/scala/scala/commit/f6056a24c5a276cd8b763992f56e122f4ddbf76d

Comment (Ichoran) [inline test/junit/scala/collection/IteratorTest.scala]: Maybe we can make it actually test iterator.  (Or discard it.)

**Comment by som-snytt** [comment]
I'll make the test changes on #4075 and delete that from this PR, since the tests have passed.  When the PR is rebased, they should pass again.  That's assuming PRs are merged forward as expected.

**Comment by som-snytt** [comment]
Repushed with readability tweaks to Iterator, but without the test files, which are updated on #4075 .

**Comment by Ichoran** [comment]
LGTM

**Comment by Ichoran** [comment]
If take, drop, and slice are implemented in terms of a template method, we need an annotation to say so when they are overridden.  Making them final is a poor substitute for warning people that there is a common implementation.  (This is hardly the only case where such an annotation would be valuable.)  Anyway, since we don't (yet?) have such an annotation, I wouldn't worry overly much about it.

**Comment by paulp** [comment]
https://twitter.com/extempore2/status/530396381651795968
https://twitter.com/extempore2/status/530396944942637057

This is especially relevant here because as I found the last time I did something about the 25 different slice implementations, there are times when you definitely want to implement slice in terms of take and drop, and there are times when you want to implement take and drop in terms of slice.

**Comment by som-snytt** [comment]
I didn't even get as far as understanding non-strict-splitter collections.

https://github.com/scala/scala/blob/2.11.x/src/library/scala/collection/parallel/RemainsIterator.scala#L408

where there are further "considerations."  It's one of those effective java words, along with "judiciously."

I want a t-shirt that says, "Consider not doing that."  Or if not a garment, then an annotation.
